### PR TITLE
Warning fix, and iexamine::atm tidy-up

### PR DESCRIPTION
**Pull request by bkentel**
Fixed a shadowed variable warning and while in the neighborhood:
- fixed up some function signatures.
- made the atm item selection filter by cash cards only.
- factored out some common code into two helpers.
- moved magic numbers to named constexprs.
- moved repeated calculations to named const variables.

### CODE REVIEW

File: src/iexamine.cpp [hunk 1]
```diff
@@ -56,39 +56,34 @@ void iexamine::gaspump(player *p, map *m, int examx, int examy)
...
-    if (p->cash >= 100) {
+    if (player_cash >= card_cost) {
         amenu.addentry( purchase_cash_card, true, -1, _("Purchase cash card?") );
     } else {
         amenu.addentry( purchase_cash_card, false, -1,
                         _("You need $1.00 in your account to purchase a card.") );
     }
 
-    if (p->has_amount("cash_card", 1) && p->cash > 0) {
+    if (card_count == 1 && player_cash > 0) {
```
Comment (narc0tiq) [inline src/iexamine.cpp]: You can only withdraw money if you have exactly one cash card?

File: src/iexamine.cpp [hunk 2]
```diff
@@ -116,164 +111,135 @@ void iexamine::atm(player *p, map *m, int examx, int examy)
...
+        if (itm.type->id != "cash_card") {
             popup(_("Please insert cash cards only!"));
+            return nullptr; // must have selected an equipped item
+        }
+
+        return &itm;
+    };
+
+    //----------------------------------------------------------------------------------------------
+    // Small helper to promp for an amount and clamp it to [0, max].
```
Comment (narc0tiq) [inline src/iexamine.cpp]: That sounds like a thing that would be useful in other places, but if it's a lambda you can only use it here.

File: src/iexamine.cpp [hunk 3]
```diff
@@ -116,164 +111,135 @@ void iexamine::atm(player *p, map *m, int examx, int examy)
...
+    if (choice == deposit_money) {
+        auto const src = choose_card(_("Insert card for deposit."));
+        if (!src) {
             return;
         }
-        if (dep->charges == 0) {
+        
+        auto const max = src->charges;
+        if (!src) {
             popup(_("You can only deposit money from charged cash cards!"));
```
Comment (narc0tiq) [inline src/iexamine.cpp]: New bug: before, it was checking if the card had any charges, now you're checking if the card exists.

File: src/iexamine.cpp [hunk 4]
```diff
@@ -116,164 +111,135 @@ void iexamine::atm(player *p, map *m, int examx, int examy)
...
-            popup(_("Please insert cash cards only!"));
+        auto const dst = choose_card(_("Insert card for bulk deposit."));
+        if (!dst) {
             return;
         }
 
-        //for all cash cards in inventory
+        // sum all cash cards in inventory
+        auto sum = dst->charges;
         for (auto &elem : g->u.inv.all_items_by_type("cash_card")) {
```
Comment (narc0tiq) [inline src/iexamine.cpp]: I know this isn't your code, but doesn't `p->charges_of("cash_card")` (used earlier) return the sum of cash card charges in the player inventory already? It seemed to be used that way.

File: src/game.h [hunk 5]
```diff
@@ -352,15 +352,15 @@ class game
...
-        int display_slice(indexed_invslice &, const std::string &, const int &position = INT_MIN);
+        int inv(const std::string &title, int position = INT_MIN);
+        int inv_activatable(std::string const &title);
+        int inv_type(std::string const &title, item_cat inv_item_type = IC_NULL);
+        int inv_for_liquid(const item &liquid, const std::string &title, bool auto_choose_single);
+        int inv_for_salvage(const std::string &title);
+        item *inv_map_for_liquid(const item &liquid, const std::string &title);
+        int inv_for_flag(const std::string &flag, const std::string &title, bool auto_choose_single);
+        int inv_for_filter(const std::string &title, item_filter filter );
+        int display_slice(indexed_invslice const&, const std::string &, int position = INT_MIN);
```
Comment (narc0tiq) [inline src/game.h]: I like the extra constness.

**Comment by narc0tiq** [comment]
I rather like this one, barring the new bugs.

**Comment by narc0tiq** [comment]
Ack. Stupid buttons right next to each other.

**Comment by bkentel** [comment]
Thanks for pointing out those small issues @narc0tiq I'll take care of them after work today.

File: src/iexamine.cpp [hunk 6]
```diff
@@ -56,39 +56,34 @@ void iexamine::gaspump(player *p, map *m, int examx, int examy)
     add_msg(m_info, _("Out of order."));
 }
 
-void iexamine::atm(player *p, map *m, int examx, int examy)
+void iexamine::atm(player *const p, map *, int , int )
 {
-    (void)m; //unused
```
Comment (DavidKeaton) [inline src/iexamine.cpp]: question: does one have to declare unused arguments as void when left out of prototype and function decl?

Comment (bkentel) [inline src/iexamine.cpp]: You can also remove their name from the parameter list, but neither are *necessary* -- they just produce a warning on most compilers. With the warning settings DDA uses though, those warnings will be flagged as errors.

Comment (KA101) [inline src/iexamine.cpp]: And errors stop the build, so yeah, voiding unused arguments is mandatory.

**Comment by kevingranade** [comment]
I like the consting, but:
Stop packing all these unrelated things into PRs, any future omnibus PRs
like this will be closed without review.
I'm serious about the superfluous lambdas, stop it.
It's unnecessary to move literals that are used once to a const.

**Comment by bkentel** [comment]
Where do you see superfluous lambdas here? These two helpers are to avoid code repetition (hence the reduction in line count despite adding vertical whitespace). How is copy and pasting code preferable to a lambda? Would it make you happier to rip those two helpers out into their own functions?

Those const values were all used at least twice I think? And they were more than simple accessory functions -- they all did some sort of internal calculation; why not cache it in a named variable? `player_cash` could/should probably be nixed though.

File: src/iexamine.cpp [hunk 7]
```diff
@@ -56,37 +56,31 @@ void iexamine::gaspump(player *p, map *m, int examx, int examy)
...
-    int pos;
-    int pos2;
-    item *dep;
-    item *with;
+    enum : int { // menu choices
+        cancel, purchase_cash_card, deposit_money,
+        withdraw_money, transfer_money, transfer_all_money
+    };
+
+    constexpr int card_cost      = 100;
```
Comment (kevingranade) [inline src/iexamine.cpp]: Did you even bother to check? card_cost and move_cost_bulk are only used one place each.

**Comment by DavidKeaton** [comment]
Late on my part, but thanks for the reply bkentel! :-) I am a supporter of -Werror but didn't know that it'd throw errors. Thanks!

File: src/iexamine.cpp (hunk 7 above)
Comment (bkentel) [inline src/iexamine.cpp]: Oh, you meant the const**expr** ones? Is it not preferable to pull out tweakable (or even jsonable) values like this?

**Comment by bkentel** [comment]
Anything pending here? @kevingranade would you like me to fold those constants back into the code?

**Comment by kevingranade** [comment]
Was there anything unclear about "It's unnecessary to move literals that are used once to a const."?
constexpr is just a stronger version of a const (it's precisely a const plus an assertion that it can be evaluated at compile-time).

**Comment by bkentel** [comment]
Making some further tweaks *atm*, hold off on merging this for the time being.

**Comment by bkentel** [comment]
+ Transformed `iexamine::atm()` to a class based implementation.
+ New: The menu no longer kicks you out after one operation.
+ New: The title displays the player's balance.
+ New: The menu accepts ESC as an option to exit from it.
+ Issue: Time will jump forward by potentially a large amount if many interactions were done. Should this be made into a player activity?

**Comment by KA101** [comment]
Long_action would be ideal, yeah, but then that applies to computer hacking too as these are functionally the same thing.

**Comment by bkentel** [comment]
Not completely sure if I've done things completely right w.r.t adding a new activity, but it seems to work as it should. If anyone could have a look, I'd appreciate it.

By the way, the automated build is failing on warning from code I didn't touch. Should be fixed regardless though.

File: src/player.h [hunk 8]
```diff
@@ -912,7 +912,7 @@ class player : public Character, public JsonSerializer, public JsonDeserializer
         int scent;
         int dodges_left, blocks_left;
         int stim, pkill, radiation;
-        unsigned long cash;
+        long cash;
```
Comment (narc0tiq) [inline src/player.h]: It's right here.

File: src/iexamine.cpp [hunk 9]
```diff
@@ -56,225 +56,278 @@ void iexamine::gaspump(player *p, map *m, int examx, int examy)
...
+        cancel, purchase_card, deposit_money, withdraw_money, transfer_money, transfer_all_money
+    };
+
+    atm_menu()                           = delete;
+    atm_menu(atm_menu const&)            = delete;
+    atm_menu(atm_menu&&)                 = delete;
+    atm_menu& operator=(atm_menu const&) = delete;
+    atm_menu& operator=(atm_menu&&)      = delete;
+
+    explicit atm_menu(player &p) : u {p} {
```
Comment (KA101) [inline src/iexamine.cpp]: Error'd in MinGW32: apparently needs to be const, not grab the player, or not from this sort of list:
C:\Documents and Settings\---\My Documents\GitHub\Cataclysm-DDA\src\iexamine.cpp|76|error: invalid initialization of non-const reference of type 'player&' from an rvalue of type '<brace-enclosed initializer list>'|

**Pull request merged**